File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-0.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<vector<int>> dp (2, vector<int>(n + 1, 0));
        dp[1][n - 1] = prices[n - 1];
        for(int i = n - 2; i > -1; --i) {
            dp[1][i] = max(prices[i] + dp[0][i + 2], dp[1][i + 1]);
            dp[0][i] = max(-prices[i] + dp[1][i + 1], dp[0][i + 1]);
        }
        return dp[0][0];
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        diameter = 0

        def traverse(node):
            nonlocal diameter
            if not node:
                return 0
            
            left_height = traverse(node.left)
            right_height = traverse(node.right)

            diameter = max(diameter, left_height + right_height)
            return 1 + max(left_height, right_height)

        traverse(root)
        return diameter